osu.Game/Online/Multiplayer/MultiplayerClient.cs

```diff
@@ -50,80 +54,64 @@ private void apiStateChanged(ValueChangedEvent<APIState> state)
-            connection.On<int>(nameof(IMultiplayerClient.HostChanged), ((IMultiplayerClient)this).HostChanged);
-            connection.On<MultiplayerRoomSettings>(nameof(IMultiplayerClient.SettingsChanged), ((IMultiplayerClient)this).SettingsChanged);
-            connection.On<int, MultiplayerUserState>(nameof(IMultiplayerClient.UserStateChanged), ((IMultiplayerClient)this).UserStateChanged);
-            connection.On(nameof(IMultiplayerClient.LoadRequested), ((IMultiplayerClient)this).LoadRequested);
-            connection.On(nameof(IMultiplayerClient.MatchStarted), ((IMultiplayerClient)this).MatchStarted);
-            connection.On(nameof(IMultiplayerClient.ResultsReady), ((IMultiplayerClient)this).ResultsReady);
-
-            connection.Closed += async ex =>
+            try
             {
-                isConnected.Value = false;
+                await disconnect(false);
 
-                Logger.Log(ex != null
-                    ? $"Multiplayer client lost connection: {ex}"
-                    : "Multiplayer client disconnected", LoggingTarget.Network);
-
-                if (connection != null)
-                    await tryUntilConnected();
-            };
-
-            await tryUntilConnected();
-
-            async Task tryUntilConnected()
-            {
-                Logger.Log("Multiplayer client connecting...", LoggingTarget.Network);
+                // this token will be valid for the scope of this connection.
+                // if cancelled, we can be sure that a disconnect or reconnect is handled elsewhere.
+                var cancellationToken = connectCancelSource.Token;
 
                 while (api.State.Value == APIState.Online)
                 {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    Logger.Log("Multiplayer client connecting...", LoggingTarget.Network);
+
                     try
                     {
-                        Debug.Assert(connection != null);
+                        // importantly, rebuild the connection each attempt to get an updated access token.
+                        connection = createConnection(cancellationToken);
 
-                        // reconnect on any failure
-                        await connection.StartAsync();
-                        Logger.Log("Multiplayer client connected!", LoggingTarget.Network);
+                        await connection.StartAsync(cancellationToken);
 
-                        // Success.
+                        Logger.Log("Multiplayer client connected!", LoggingTarget.Network);
                         isConnected.Value = true;
-                        break;
+                        return;
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        //connection process was cancelled.
+                        throw;
```

Comment:
I'm not super sure of this re-throw, especially given that we use cancellation to clean up in disposal now. Feels like this might end up killing the game unintentionally instead of just cancelling the operation. Maybe it should be a return?

Reply:
This is how you're supposed to handle `OperationCancelledExceptions`. I mean we can silently hide it here and break conventions if you really think that's beneficial, but I'd rather code this in a way where it can be used in proper async code in the future, if/when required.

Reply:
I somehow missed that these were caught. My bad.